models/base.py

```diff
@@ -167,7 +168,72 @@ def SmartSession(*args):
+            # later cause database deadlocks because of the table locks we
+            # use.  It's probably depending on garbage collection, and
+            # sometimes the garbage doesn't get collected in time.  So,
+            # explicitly close and invalidate the session.
+            #
+            # NOTE -- this doesn't seem to have actually fixed the problem. :(
+            # I've tried to hack around it by putting a timeout on the locks
+            # with a retry loop.  Sigh.
+            #
+            # Even *that* doesn't seem to have fully fixed it.
+            # *Sometimes*, not reproducibly, there's a session that
+            # hangs around that is idle in transaction.  There must be
+            # some reference to it *somewhere* that's stopping it from
+            # getting garbage collected.  I really wish SQLA just closed
+            # the connection when I told it to.  I tried adding
+            # "session.rollback()" here, but then got all kinds of
+            # deatched instance errors trying to access objects later.
+            # It seems that rollback() subverts the session's
+            # expire_on_commit=False setting.
+            #
+            # OOO, ooo, here's an idea: just use SQL to rollback.  Hopefully
+            # SQLAlchemy won't realize what we're doing and won't totally
+            # undermine us for doing it.  (My god I hate SQLA.)
+            # (What I'm really trying to accomplish here is given that we
+            # seem to rarely have an idle session sitting around, make sure
+            # it's not in a transaction that will prevent table locks.)
+            #
+            # session.execute( sa.text( "ROLLBACK" ) )
+            #
+            # NOPE!  That didn't work.  If there was a previous
+            # exception, sqlalchemy catches that before it lets me run
+            # session.execute, saying I gotta rollback before doing
+            # anything else.  (There is irony here.)
+            #
+            # OK, lets try grabbing the connection from the session and
+            # manually rolling back with psycopg2 or whatever is
+            # underneath.  I'm not sure this will do what I want either,
+            # because I don't know if session.bind.raw_connection() gets
+            # me the connection that session is using, or if it gets
+            # another connection.  (If the latter, than this code is
+            # wholly gratuitous.)
+            #
+            # dbcon = session.bind.raw_connection()
+            # cursor = dbcon.cursor()
+            # cursor.execute( "ROLLBACK" )
+
+            # ...even that doesn't seem to be solving the problem.
+            # The solution may end up being moving totally away from
+            # SQLAlchemy and using something that lets us actually
+            # control our database connections.
+
+            # OK, another thing to try.  See if expunging all objects
+            # lets me rollback.
```

Comment:
This comment chain was a journey to experience... I felt like I was right there with you!